src/cryptography/hazmat/primitives/kdf/concatkdf.py

```diff
@@ -71,36 +81,46 @@ def __init__(self, algorithm, length, otherinfo, backend=None):
         self._backend = backend
         self._used = False
 
-    def _hash(self):
+    def _hash(self) -> hashes.Hash:
         return hashes.Hash(self._algorithm, self._backend)
 
-    def derive(self, key_material):
+    def derive(self, key_material: bytes) -> bytes:
         if self._used:
             raise AlreadyFinalized
         self._used = True
         return _concatkdf_derive(
             key_material, self._length, self._hash, self._otherinfo
         )
 
-    def verify(self, key_material, expected_key):
+    def verify(self, key_material: bytes, expected_key: bytes) -> None:
         if not constant_time.bytes_eq(self.derive(key_material), expected_key):
             raise InvalidKey
 
 
-@utils.register_interface(KeyDerivationFunction)
-class ConcatKDFHMAC(object):
-    def __init__(self, algorithm, length, salt, otherinfo, backend=None):
+class ConcatKDFHMAC(KeyDerivationFunction):
+    def __init__(
+        self,
+        algorithm: hashes.HashAlgorithm,
+        length: int,
+        salt: typing.Optional[bytes],
+        otherinfo: typing.Optional[bytes],
+        backend=None,
+    ):
         backend = _get_backend(backend)
 
         _common_args_checks(algorithm, length, otherinfo)
         self._algorithm = algorithm
         self._length = length
-        self._otherinfo = otherinfo
-        if self._otherinfo is None:
-            self._otherinfo = b""
+        self._otherinfo: bytes = otherinfo if otherinfo is not None else b""
+
+        block_size = getattr(algorithm, "block_size", None)
```

Comment:
Hmm?

Reply:
`block_size` isn't a part of the `HashAlgorithm` interface (and we don't publicly document it). Adding it arguably wouldn't make sense either since sponge constructions like SHA3 don't have an internal block size either. mypy made this bug obvious so I added a new branch and test to make sure only hashes with `block_size` could be used.

I'm open to other solutions here.

Reply:
`BlockHashAlgorithm`?

Reply:
Would need to be a subclass of `HashAlgorithm` for mypy to allow it because mypy doesn't have intersection types (so you can't have a multiply-inherited class that is both `HashAlgorithm` and `BlockHashAlgorithm`).

Reply:
That seems reasonable?

Reply:
Are we comfortable with changing our CipherAlgorithms in that fashion as well? (e.g. `BlockCipherAlgorithm` is a subclass of `CipherAlgorithm`, whereas now it is not). If so, then okay, I'll do that here too.

Reply:
Also, would we want to document these new properties in the docs?

Reply:
I'm... not sure. Thoughts?

Reply:
If they're exposed as part of the typing it seems like we need to document them. I guess it matters in this case since ConcatKDF would literally have failed with a TypeError (when it tried to multiply a `None`) if you passed SHA3 to it.

Reply:
If `ConcatKDF` has the property, but it's simply `None`, maybe we should add that to the base interface as `Optional[int]`, then simply check it's not none in the places we use it. That seems better than the `getattr` shenangins.

Reply:
We can do that here, yes. I'd be tempted to not document that property on `HashAlgorithm` though, since it's not generally useful to our users.

Reply:
That's fine with me.

Would you mind doing that bit in it's own PR for ease of review? I'm trying to push us towards the typing PRs not introducing substantive changes themselves.

Reply:
I had no intention of doing it in this PR. This is already a 290 line diff :)